### PR TITLE
Kraken wsCancelAllOrders added, fix bugs in wsAddOrder

### PR DESCRIPTION
**Pull request by vazha**
# PR Description

Fixed some Kraken ws issues, added method,

Fixes # (issue)

Kraken wsAddOrder requires price and volume in string format. Also fixed pair formatting for it. Fixed exceeded timeout response.
Added wsCancelAllOrders for Kraken.
Kraken wsCancelOrders - fixed exceeded timeout response.
Kraken wsProcessOwnTrades - fixed saving in orders storage.

## Type of change

- [x] Bug fix (non-breaking change which fixes an issue)

## How has this been tested

- [x] go test ./... -race

## Checklist

- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas

### CODE REVIEW

File: exchanges/kraken/kraken_wrapper.go [hunk 1]
```diff
@@ -634,6 +642,16 @@ func (k *Kraken) CancelAllOrders(_ *order.Cancel) (order.CancelAllResponse, erro
 		Status: make(map[string]string),
 	}
 
+	if k.Websocket.CanUseAuthenticatedWebsocketForWrapper() {
+		resp, err := k.wsCancelAllOrders()
+		if err != nil {
+			return cancelAllOrdersResponse, err
+		}
+
+		cancelAllOrdersResponse.Status["Count"] = fmt.Sprint(resp.Count)
```
Comment (gloriousCode) [inline exchanges/kraken/kraken_wrapper.go:651]: So the purpose of the status map is supposed to be that you get orderIDs and their status eg `["133713371337"]"Cancelled"`
I think it would be better to update `CancelAllResponse` struct with a `Count` int64 field and set that instead.

File: testdata/configtest.json [hunk 2]
```diff
@@ -1,5 +1,6 @@
 {
```
Comment (gloriousCode) [inline testdata/configtest.json:1]: Please revert these changes as we use `configtest` to check config change compatibility

**Review by gloriousCode — CHANGES_REQUESTED** [review]
Very nice @vazha ! This looks really good. Just a couple of comments from me

File: exchanges/kraken/kraken_wrapper.go [hunk 3]
```diff
@@ -815,7 +839,7 @@ func (k *Kraken) GetActiveOrders(req *order.GetOrdersRequest) ([]order.Detail, e
 		return nil, err
 	}
 
-	fmt, err := k.GetPairFormat(asset.Spot, true)
+	format, err := k.GetPairFormat(asset.Spot, true)
```
Comment (shazbert) [inline exchanges/kraken/kraken_wrapper.go:842]: can you please change asset.Spot to assetType

File: exchanges/kraken/kraken_wrapper.go [hunk 4]
```diff
@@ -870,7 +894,7 @@ func (k *Kraken) GetOrderHistory(getOrdersRequest *order.GetOrdersRequest) ([]or
...
```
Comment (shazbert) [inline exchanges/kraken/kraken_wrapper.go:897]: asset.Spot -> getOrdersRequest.AssetType

File: engine/rpcserver.go [hunk 5]
```diff
@@ -1021,7 +1021,23 @@ func (s *RPCServer) CancelOrder(_ context.Context, r *gctrpc.CancelOrderRequest)
...
+	if r.Exchange != "kraken" { // at the moment implemented only on the Kraken
+		return &gctrpc.CancelAllOrdersResponse{}, common.ErrNotYetImplemented
+	}
+
+	exch := s.GetExchangeByName(r.Exchange)
+	if exch == nil {
+		return &gctrpc.CancelAllOrdersResponse{}, errExchangeNotLoaded
+	}
+
+	resp, err := exch.CancelAllOrders(nil)
```
Comment (shazbert) [inline engine/rpcserver.go:1033]: Some exchanges like Bithumb, BTSE utilise order.Cancel to retrieve open orders depending on order-side and currencies. This needs to be reflected and options added to the RPC and GCTCLI.

File: engine/rpcserver.go [hunk 6]
```diff
@@ -1021,7 +1021,23 @@ func (s *RPCServer) CancelOrder(_ context.Context, r *gctrpc.CancelOrderRequest)
...
+	if err != nil {
+		return &gctrpc.CancelAllOrdersResponse{}, err
+	}
+
+	return &gctrpc.CancelAllOrdersResponse{
+		Count: resp.Status["Count"], // count of deleted orders
```
Comment (shazbert) [inline engine/rpcserver.go:1039]: There are some instances that this map will be nil. See CoinbasePro so a nil check should be done prior to accessing the map.

**Review by shazbert — CHANGES_REQUESTED** [review]
Looking good, just a few issues that I found in conjunction with what GloriousCode suggested.

File: exchanges/kraken/kraken_wrapper.go [hunk 7]
```diff
@@ -621,7 +622,14 @@ func (k *Kraken) CancelOrder(o *order.Cancel) error {
 		return err
 	}
 	if k.Websocket.CanUseAuthenticatedWebsocketForWrapper() {
-		return k.wsCancelOrders([]string{o.ID})
+		var ordersList []string
```
Comment (MadCozBadd) [inline exchanges/kraken/kraken_wrapper.go:625]: CancelOrder(o *order.Cancel) is primarily assumes singular orderIDs will be passed in. If we start passing in orderIds with comma separated string for some exchanges and not for others, then that becomes an issue. Solution for this would be to add CancelBatchOrders to the wrappers. This would be a big task but to make it easier you could handle kraken properly and for the rest of the exchanges just return nil, common.ErrFunctionNotImplemented

**Review by MadCozBadd — COMMENTED** [review]
Thanks for all ur contributions m88888888

File: exchanges/kraken/kraken.go [hunk 8]
```diff
@@ -55,6 +55,9 @@ const (
 	// Rate limit consts
 	krakenRateInterval = time.Second
 	krakenRequestRate  = 1
+
+	// Status consts
+	Open = "open"
```
[the remark this answers is not in the document]
Comment (MadCozBadd) [reply exchanges/kraken/kraken.go:60]: Open -> open (i mightve highlighted the incorrect line)

File: exchanges/kraken/kraken_wrapper.go [hunk 9]
```diff
@@ -621,7 +622,14 @@ func (k *Kraken) CancelOrder(o *order.Cancel) error {
...
+		orders := strings.Split(o.ID, ",")
+		for _, orderID := range orders {
+			ordersList = append(ordersList, orderID)
+		}
+
+		err := k.wsCancelOrders(ordersList)
+		return err
```
[the remark this answers is not in the document]
Comment (vazha) [reply exchanges/kraken/kraken_wrapper.go:632]: :+1:

File: exchanges/kraken/kraken_wrapper.go (hunk 7 above)
Comment (vazha) [reply exchanges/kraken/kraken_wrapper.go:625]: done
